Реализованы добавление пользователей в друзья и возможность поставить лайк фильму

Natal1a-Chuklina opened the pull request.

Natal1a-Chuklina committed:
[+] ErrorHandler.class - добавлен класс для централизованной обработки ошибок
[r] FilmController.class - логика работы приложения и хранение данных вынесены в отдельные классы
[+] FilmController.class - добавлены эндпоинты для получения фильма по id, добавления и удаления лайков фильма, получения наиболее популярных фильмов
[r] UserController.class - логика работы приложения и хранение данных вынесены в отдельные классы
[+] UserController.class - добавлены эндпоинты для получения пользователя по id, добавления и удаления пользователей из друзей, получения списка друзей пользователя и получения списка общих друзей
[+] ErrorResponse.class - добавлен класс для возврата сообщений об ошибках
[+] Film.class - добавлено поле для хранения id пользователей, которые поставили лайк фильму
[+] User.class - добавлено поле для хранения id друзей пользователя
[+] FilmService.class - добавлен класс с логикой работы приложения c фильмами
[+] UserService.class - добавлен класс с логикой работы приложения с пользователями
[+] FilmStorage.interface - добавлен интерфейс для классов, отвечающих за хранение фильмов
[+] UserStorage.interface - добавлен интерфейс для классов, отвечающих за хранение пользователей
[+] InMemoryFilmStorage.class - добавлен класс для хранения фильмов в памяти приложения
[+] InMemoryUserStorage.class - добавлен класс для хранения пользователей в памяти приложения
[+] Constants.class - добавлен класс для хранения констант
[s] FilmorateApplication.class - косметические правки
[+] FilmServiceTest - добавлены тесты для проверки корректности работы приложения с фильмами
[+] UserServiceTest - добавлены тесты для проверки корректности работы приложения с пользователями

z8q commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,52 +1,62 @@
+import java.util.List;
 
 @RestController
 @Slf4j
```
Аннотацию оставили, а логи убрали.
Но на самом деле хорошо использовать логирование на уровне контроллеров.
Мы сообщаем о том, что запрос пришел и начинаем выполнение, пример:
`log.info("Creating film {}", film);`
В случае, если будут падения, мы добавим в лог сообщение об ошибке. В случае успешного выполнения добавим в лог информацию об успехе.

Natal1a-Chuklina replied:
Логи добавила

z8q commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,52 +1,62 @@
+
+    private final FilmService filmService;
+
+    public FilmController(FilmService filmService) {
```
Будет ли работать без `@Autowired`? На втором контроллере тоже нужно посмотреть

Natal1a-Chuklina replied:
Аннотация не нужна, когда в классе только один конструктор. Все работает, иначе тесты бы не прошли.

z8q commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,106 @@
+    private final FilmStorage filmStorage;
+    private final UserStorage userStorage;
+
+    public FilmService(FilmStorage filmStorage, UserStorage userStorage) {
```
Тоже не хватает `@Autowired`

z8q commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -25,11 +27,21 @@ public class Film {
     private LocalDate releaseDate;
     @Positive
     private int duration;
+    private final List<Integer> likes;
```
Можно было бы тут и в User использовать Set в качестве коллекции, чтобы избежать повторов

Natal1a-Chuklina replied:
Коллекции поменяла

z8q commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,132 @@
+        }
+
+        log.info("Получен список общих друзей пользователей с id: {} и {}", userId, otherId);
+        return userStorage.getUser(userId).getFriends().stream().filter(friendId -> userStorage.getUser(otherId)
```
Если будешь использовать Set для хранения списка друзей, то можно рассмотреть использование метода у HashSet – retainAll() для выборки общих друзей

Natal1a-Chuklina replied:
Оставила стримы

z8q commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,132 @@
+    private int counter = 1;
+    private final UserStorage userStorage;
+
+    public UserService(UserStorage userStorage) {
```
`@Autowired`

z8q reviewed (changes requested):
Привет!
Комментарии оставил)

Natal1a-Chuklina committed:
[+] FilmController.class - добавлены логи
[+] UserController.class - добавлены логи
[r] Film.class - коллекция для хранения id лайкнувших фильм пользователей изменена на HashSet
[r] User.class - коллекция для хранения id друзей изменена на HashSet
[r] FilmService.class - изменен код класса с учетом изменения типа коллекции в Film
[r] UserService.class - изменен код класса с учетом изменения типа коллекции в User
[s] StartDateValidatorTest.class - косметические правки
[r] FilmServiceTest - изменены тесты с учетом изменения типа коллекции в Film
[r] UserServiceTest - изменены тесты с учетом изменения типа коллекции в User
